EE-438: get_caller can be used in all code

birchmd opened the pull request:
### Overview
Allows `get_caller` to be used in session code, which in turn will allow a any code to look up what account it is running in, and this will be needed in writing the `main_purse()` function. The `main_purse` function itself is needed to make token transfers (including to the PoS contract) possible.

### Which JIRA ticket does this PR relate to?
https://casperlabs.atlassian.net/browse/EE-438

### Complete this checklist before you submit this PR
- [x] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [CasperLabs coding standards](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/16842753/Coding+Standards).
- [x] If this PR adds a new feature, it includes tests related to this feature.
- [x] You assigned one person to review this PR.
- [x] Your GitHub account is linked with our [Drone CI](http://drone.casperlabs.io/) system. This is necessary to run tests on this PR.

### Notes
_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._

EdHastingsCasperLabs commented on execution-engine/engine/src/execution.rs:
```diff
@@ -1013,8 +1008,8 @@ where
-                self.get_caller(dest_ptr)
-                    .map(|status| Some(RuntimeValue::I32(status)))
+                self.get_caller(dest_ptr)?;
+                Ok(None)
```
not sure if I like Ok(None).

EdHastingsCasperLabs commented on execution-engine/engine/src/execution.rs:
```diff
@@ -1013,8 +1008,8 @@ where
-                self.get_caller(dest_ptr)
-                    .map(|status| Some(RuntimeValue::I32(status)))
+                self.get_caller(dest_ptr)?;
+                Ok(None)
```
This feels problematic to me. Last week there was discussion about 0 being success. And panic vs propagating seems wrong. why not return a non-zero value for args parse failure and a different non zero value for get_caller failure?

henrytill replied:
ohhai (see below)

birchmd replied:
There is no recoverable error here. If `get_caller` failed then it is because of something really low level, like wasmi failed to write to the memory or something. I think causing a `Trap` is the correct course of action if there is any error.

I agree that returning status codes (or other return value) is a good practice when you expect the other side to be able to make use of the information, but I do not think that is the case here.

henrytill commented on execution-engine/common/src/lib.rs:
```diff
@@ -90,7 +90,7 @@ mod ext_ffi {
         pub fn set_action_threshold(permission_level: u32, threshold: i32) -> i32;
         pub fn remove_uref(name_ptr: *const u8, name_size: usize);
-        pub fn get_caller(dest_ptr: *const u8) -> i32;
+        pub fn get_caller(dest_ptr: *const u8);
```
I feel a bit uneasy about this pattern of not using a return code/value for functions like this.  Seems better suited to functions that truly have a unit return type (like `remove_uref`). 

I don't think it warrants changing yr current PR, but I do think we need to establish a set of guidelines and best practices for defining these functions that will be called from the FFI.

birchmd replied:
Totally happy to talk about guidelines for sure. I gave my reasoning here in my reply to Ed https://github.com/CasperLabs/CasperLabs/pull/747#discussion_r299695804

But we can circle back and have a larger discussion about this at a later date.

birchmd commented:
bors r+

EdHastingsCasperLabs commented on execution-engine/engine/src/execution.rs:
```diff
@@ -313,17 +313,12 @@ where
     /// If caller is defined (it's a subcall) then writes caller public key
     /// to [dest_ptr] in the Wasm memory and returns 1.
```
does the comment need to get changed to remove the subcall mention ?

birchmd replied:
Ah, yes it should, good catch.

birchmd commented:
Sorry for the force push. I addressed your last comment about updating the docstring @EdHastingsCasperLabs (didn't feel like that deserved its own commit) and fixed a typo in a previous commit message.

birchmd commented:
bors retry

birchmd commented:
bors r-

birchmd commented:
bors r+
